components/Router.coffee

```diff
@@ -1,11 +1,137 @@
+uuid = require 'uuid'
 
-exports.getComponent = (metadata) ->
-  component = BaseRouter.getComponent()
+validVerbs = ['all', 'get', 'post', 'put', 'delete', 'options', 'patch', 'head']
 
-  noflo.helpers.WirePattern component,
-    in: ['app']
-    out: []
-    params: ['pattern']
-  , BaseRouter.getProcess component, false, false
+exports.getComponent = ->
+  c = new noflo.Component
+    description: 'Static index-based Express router'
+    inPorts:
+      app:
+        datatype: 'object'
+        type: 'http://expressjs.com/4x/api.html#app'
+        description: 'Express Application or Router'
+        required: true
+        control: true
+      pattern:
+        datatype: 'string'
+        description: 'Request patterns as `verb /path`
+        or just `/path` meaning verb is `all`'
+        addressable: true
+        required: true
+      path:
+        datatype: 'string'
+        description: 'Restrict this branch to a specific URL /root'
+        required: true
+        control: true
+        default: '/'
+      filters:
+        datatype: 'array'
+        description: 'Route filter middleware (omitted by default)'
+        required: true
+        control: true
+    outPorts:
+      req:
+        datatype: 'object'
+        type: 'http://expressjs.com/4x/api.html#req'
+        description: 'Express Request objects (contain responses)'
+        addressable: true
+      error:
+        datatype: 'object'
+        required: false
+
+  c.process (input, output) ->
+    # scoped variables for conditionals
+    hasPath = false
+    hasFilter = false
+
+    # precondition
+    return unless input.has 'app'
+
+    # if attached, it has a filter
+    if input.ports.filters.isAttached()
+      return unless input.has 'filters'
+      hasFilter = true
+      filters = input.getData 'filters'
+
+    # if attached, it has a path
+    if input.ports.path.isAttached()
+      return unless input.has 'path'
+      rootPath = input.getData 'path'
+      hasPath = true
+
+    # process...
+    patterns = []
+    handlers = []
+    app = input.getData 'app'
+    pattern = input.buffer.get 'pattern'
+      .filter (ip) -> ip.type is 'data'
+      .map (ip) -> ip.data
+
+    # put the pattern input array<string>s into objects
+    for index, pat of pattern
+      pat = pat.split /\s+/
+      verb = if pat.length is 2 then pat[0] else 'all'
+      path = if pat.length is 2 then pat[1] else pat[0]
+      unless validVerbs.indexOf(verb) >= 0
+        return output.done new Error "Invalid HTTP verb: '#{verb}'"
+      unless path
+        return output.done new Error "Incorrect HTTP path: '#{path}'"
+      patterns[Number index] =
+        verb: verb
+        path: path
+
+    router = express.Router()
+
+    # Adding the routes here
+    for pat, index in patterns
+      do (pat, index) ->
+        return if pat is undefined
+        handlers[index] = (req, res, next) ->
+          openBracket = new noflo.IP 'openBracket', req.uuid, index: index
```

Comment:
I was planning to use scopes for request isolation here, not brackets. We should really stop using brackets for anything other than streams.